resolve.go

```diff
@@ -9,59 +9,81 @@ import (
 )
 
 var ResolvableProtocols = []ma.Protocol{DnsaddrProtocol, Dns4Protocol, Dns6Protocol, DnsProtocol}
-var DefaultResolver = &Resolver{Backend: net.DefaultResolver}
+var DefaultResolver = &Resolver{def: net.DefaultResolver}
 
 const dnsaddrTXTPrefix = "dnsaddr="
 
-type Backend interface {
+// BasicResolver is a low level interface for DNS resolution
+type BasicResolver interface {
 	LookupIPAddr(context.Context, string) ([]net.IPAddr, error)
 	LookupTXT(context.Context, string) ([]string, error)
 }
 
+// Resolver is an object capable of resolving dns multiaddrs by using one or more BasicResolvers;
+// it supports custom per domain/TLD resolvers.
+// It also implements the BasicResolver interface so that it can act as a custom per domain/TLD
+// resolver.
 type Resolver struct {
-	Backend Backend
```

Comment:
@vyzo IIRC this is making a choice to break backwards compatibility here a bit for the sake of future extensibility.

Two breakages:
- Rename `Backend` to `BasicResolver`, both not perfect names but `Backend` is probably worse.
   - We're not really gaining much out of the name change, but it also seems low cost for people to update. If we really wanted to we could just alias `type Backend BasicResolver` and deprecate it, but idk if it's worth bothering
- Constructing `Resolver` is now different and we can't do `Resolver.Backend` anymore to get out the `BasicResolver`
   - Neither of these seems like a big deal and moving to a construction with `Options` gives us more flexibility to add more options in the future

cc @Stebalien in case you have any issues with the breakages.

Reply:
Yeah, conscious choice -- having the fields public was just a bad idea.
I think the original emerged in a hacky way when the ability to change the backend was added.